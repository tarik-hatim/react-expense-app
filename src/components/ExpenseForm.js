import React from 'react';
import 'react-dates/initialize';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
// import {apiService} from '../api/apiService';
// console.log(apiService);


const now = moment();
now.locale('fr', {
    months : 'janvier_février_mars_avril_mai_juin_juillet_août_septembre_octobre_novembre_décembre'.split('_'),
    monthsShort : 'janv._févr._mars_avr._mai_juin_juil._août_sept._oct._nov._déc.'.split('_'),
    monthsParseExact : true,
    weekdays : 'dimanche_lundi_mardi_mercredi_jeudi_vendredi_samedi'.split('_'),
    weekdaysShort : 'dim._lun._mar._mer._jeu._ven._sam.'.split('_'),
    weekdaysMin : 'Di_Lu_Ma_Me_Je_Ve_Sa'.split('_'),
    weekdaysParseExact : true,
    longDateFormat : {
        LT : 'HH:mm',
        LTS : 'HH:mm:ss',
        L : 'DD/MM/YYYY',
        LL : 'D MMMM YYYY',
        LLL : 'D MMMM YYYY HH:mm',
        LLLL : 'dddd D MMMM YYYY HH:mm'
    },
    calendar : {
        sameDay : '[Aujourd’hui à] LT',
        nextDay : '[Demain à] LT',
        nextWeek : 'dddd [à] LT',
        lastDay : '[Hier à] LT',
        lastWeek : 'dddd [dernier à] LT',
        sameElse : 'L'
    },
    relativeTime : {
        future : 'dans %s',
        past : 'il y a %s',
        s : 'quelques secondes',
        m : 'une minute',
        mm : '%d minutes',
        h : 'une heure',
        hh : '%d heures',
        d : 'un jour',
        dd : '%d jours',
        M : 'un mois',
        MM : '%d mois',
        y : 'un an',
        yy : '%d ans'
    },
    dayOfMonthOrdinalParse : /\d{1,2}(er|e)/,
    ordinal : function (number) {
        return number + (number === 1 ? 'er' : 'e');
    },
    meridiemParse : /PD|MD/,
    isPM : function (input) {
        return input.charAt(0) === 'M';
    },
    // In case the meridiem units are not separated around 12, then implement
    // this function (look at locale/id.js for an example).
    // meridiemHour : function (hour, meridiem) {
    //     return /* 0-23 hour, given meridiem token and hour 1-12 */ ;
    // },
    meridiem : function (hours, minutes, isLower) {
        return hours < 12 ? 'PD' : 'MD';
    },
    week : {
        dow : 1, // Monday is the first day of the week.
        doy : 4  // Used to determine first week of the year.
    }
});
//console.log(now.format("D, MMMM YYYY, h:mm a"));
export default class ExpenseForm extends React.Component{

    
    state = {
        description: this.props.expense ? this.props.expense.description : '',
        note: this.props.expense ? this.props.expense.note :'',
        amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
        createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
        focused: false,
        error: ''
        
    }

    onDescriptionChange = (e)=> {
        const description = e.target.value;
        this.setState(()=>({ description}))
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(()=> ({ note }));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
       
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() =>({ amount }))
        }
    }
    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({ createdAt }))
        }
        
    }

    onFocusChange = ({ focused}) => {
        this.setState(() => ({ focused }))
    }
    onSubmit = (e) => {
        e.preventDefault();
       if(!this.state.description || !this.state.amount){
           const error = 'Please provide description and amount.';
           this.setState((()=>({
               error
           })))
       }else{
           this.setState(() => ({
                error : ''
           }));
           this.props.onSubmit({
                description : this.state.description,
                amount: parseFloat(this.state.amount,10) * 100,
                createdAt : this.state.createdAt.valueOf(),
                note: this.state.note
           });
       }
        
    }
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text" placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}

                    />
                    <input
                        type="text" placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <textarea placeholder="add a note for the expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}>
                    </textarea>
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={this.onDateChange} // PropTypes.func.isRequired
                        focused={this.state.focused} // PropTypes.bool
                        onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                        id="your_unique_id" // PropTypes.string.isRequired,
                        numberOfMonths={1}
                        isOutsideRange={() => {false}}
                    />
                    <button>Add Expense</button>
                </form>
            </div>
        )
        
    }
}

