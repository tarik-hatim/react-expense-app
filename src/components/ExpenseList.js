import React from 'react';
import { connect } from 'react-redux';
import ExpenseItem from './ExpenseListItem';
import visibleExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        <p>expense List </p>
        {props.expenses.length === 0 ? (
            <p>no expenses</p>
        ): (
            props.expenses.map((expense) => 
            <ExpenseItem  {...expense} key={expense.id} /> ))}
    </div>
);

const mapStateToProps = (state) => (
    {
        expenses : visibleExpenses(state.expenses,state.filters)
    }
);

export default connect(mapStateToProps)(ExpenseList);
