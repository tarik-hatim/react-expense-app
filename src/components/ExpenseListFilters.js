import React from 'react';
import { connect } from 'react-redux';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setEndDate, setStartDate } from '../actions/filters';


class ExpenseListFilters extends React.Component{
    constructor(props){
        super(props);
        console.log(props);

        let focusedInput = null;
        if (props.autoFocus) {
            focusedInput = START_DATE;
        } else if (props.autoFocusEndDate) {
            focusedInput = END_DATE;
        }

        this.state = {
            focusedInput : null
        
        }
    }

    onDatesChange = ({startDate, endDate}) => {
        
        this.props.dispatch(setStartDate(startDate))
        this.props.dispatch(setEndDate(endDate))
    }

    onFocusChange = (focusedInput) =>{

        this.setState({ focusedInput});
    }

    render(){
        return (

            <div>
                <input type="text" value={ this.props.filters.text } onChange={(e) =>{ 
                this.props.dispatch(setTextFilter(e.target.value))
                }} />
                <select value={this.props.filters.sortBy} onChange={(e) => {
                    console.log(e.target.value);
                    if(e.target.value === 'date'){
                        this.props.dispatch(sortByDate());
                    }else if(e.target.value === 'amount'){
                        this.props.dispatch(sortByAmount());
                    }
                }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                    
                    endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                    startDateId="start_date"
                    endDateId="end_date"
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                    isOutsideRange={()=>{ false}}
                    numberOfMonths={1}
                    showClearDates={true}
                />
                
            </div>
        );
    }
}

// const ExpenseListFilters = (props) => (
//     <div>
//         <input type="text" value={ props.filters.text } onChange={(e) =>{
            
//             props.dispatch(setTextFilter(e.target.value))
//         }} />
//         <select value={props.filters.sortBy} onChange={(e) => {
//             console.log(e.target.value);
//             if(e.target.value === 'date'){
//                 props.dispatch(sortByDate());
//             }else if(e.target.value === 'amount'){
//                 props.dispatch(sortByAmount());
//             }
//         }}>
//             <option value="date">Date</option>
//             <option value="amount">Amount</option>
//         </select>
//         <select value={props.filters.startDate}></select>
//     </div>
// );

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

export default connect(mapStateToProps)(ExpenseListFilters);