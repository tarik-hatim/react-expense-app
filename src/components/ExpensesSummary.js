 import React from 'react';
 import { connect } from 'react-redux';
 import visibleExpenses from '../selectors/expenses';
 import expensesTotal from '../selectors/expenses-total';
 import numeral from 'numeral';

 export const ExpensesSummary = ({expensesCount, expensesTotal}) =>  {
     const expenseWord = expensesCount > 1 ? 'expenses ' : 'expense ';
     const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <div>
           <h1>viewing {expensesCount} {expenseWord} totalling {formattedExpensesTotal}</h1>
        </div>
    )
 }

 const mapStateToProps = (state) => (
    {
        expensesTotal : expensesTotal(visibleExpenses(state.expenses,state.filters)),
        expensesCount: (visibleExpenses(state.expenses, state.filters)).length
    }
 )
    
 

 export default connect(mapStateToProps)(ExpensesSummary);