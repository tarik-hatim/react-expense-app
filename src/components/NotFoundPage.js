import React from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
    <div>
    <h2>
        No match found go to <Link to="/">home page</Link>
    </h2>
    </div>
);

export default NotFoundPage;