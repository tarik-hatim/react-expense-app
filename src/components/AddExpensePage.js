import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import { addExpense} from '../actions/expenses';

 export class  AddExpensePage extends React.Component{

  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  }
  render(){
    return (
      <div>
        <h2>Add Expense Page</h2>
        <ExpenseForm onSubmit={this.onSubmit} />
    </div>
    )
  }
    
}

const mapDispatchToProps = dispatch =>  {
  return {
    addExpense : (expense) => dispatch(addExpense(expense)),
    
  }

};

export default connect(null,mapDispatchToProps)(AddExpensePage);