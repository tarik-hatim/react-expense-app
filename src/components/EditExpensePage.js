import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';


export class EditExpensePage extends React.Component{


  onSubmit = (expense) => {
      this.props.editExpense(this.props.expense.id,expense);
      this.props.history.push('/');
  }
  onRemove = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  }
  render(){
    return (
      <div>
        <h2>Edit Expense Page</h2>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}>remove</button>
      </div>
    );
  }
}


const mapStateToProps = (state, props) => {
//console.log(props.match.params.id);
    return {
      expense : state.expenses.find((expense) => expense.id === props.match.params.id)  
    };
};

const mapDispatchToProps = props => {
  return {
    editExpense : (expense) => props.dispatch(editExpense(props.expense.id, expense)),
    removeExpense : (data) => props.dispatch(removeExpense({ id: props.expense.id }))
  }
  
}

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);