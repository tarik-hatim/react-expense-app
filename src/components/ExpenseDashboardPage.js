import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters  from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage =() => (
    <div>
      <ExpensesSummary />
      <ExpenseListFilters />
      <h2>Dashboard</h2>
      <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;