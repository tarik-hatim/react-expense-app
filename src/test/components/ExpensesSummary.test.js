import React from 'react';
import { shallow } from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';

test('Should render ExpensesSummary with one expense', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={50455} />);
    expect(wrapper).toMatchSnapshot();
});
test('Should render ExpensesSummary with expenses', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={3} expensesTotal={750854} />);
    expect(wrapper).toMatchSnapshot();
});