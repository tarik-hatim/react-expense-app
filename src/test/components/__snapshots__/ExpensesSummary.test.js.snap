// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary with expenses 1`] = `
<div>
  <h1>
    viewing 
    3
     
    expenses 
     totalling 
    $7,508.54
  </h1>
</div>
`;

exports[`Should render ExpensesSummary with one expense 1`] = `
<div>
  <h1>
    viewing 
    1
     
    expense 
     totalling 
    $504.55
  </h1>
</div>
`;
